### ISSUE TITLE
prove predictive risk and hotspot models are trained, versioned, and deployed correctly

### ISSUE THREAD

**Issue by Aadhithya-balu** [Aadhithya-balu/Saksha_Datathon#165]
- Scope:
  - Validate the model artifact pipeline and metadata generation.
  - Confirm model and feature columns match the actual input schema.
  - Require real training artifacts before claiming predictive intelligence is live.
- Acceptance:
  - Every prediction endpoint exposes model provenance and loaded status.
  - Production deployment blocks if model artifacts are missing or mismatched.
- Relevant files: `backend/app/ai/inference/risk.py`, `backend/app/ai/inference/hotspot.py`, `backend/app/ai/models/*`

## PROMPT

```
````text
# Issue — Validate ML Model Artifacts, Feature Schema, and Production Readiness

## Objective

Strengthen the SAKSHA predictive intelligence pipeline so that every risk/hotspot prediction is generated only from a valid, compatible, and explicitly identified model artifact.

Relevant areas:

- `backend/app/ai/inference/risk.py`
- `backend/app/ai/inference/hotspot.py`
- `backend/app/ai/models/*`

The current system contains real inference logic and fallback behavior, but predictive intelligence must not be considered "live" unless the required trained model artifacts are actually available, valid, compatible with the current feature schema, and successfully loaded.

The goal is to make model provenance, artifact validation, and feature compatibility explicit and production-safe.

---

# 1. AUDIT THE EXISTING MODEL PIPELINE

First inspect the complete predictive pipeline.

Review:

- `backend/app/ai/inference/risk.py`
- `backend/app/ai/inference/hotspot.py`
- `backend/app/ai/models/*`
- Model loading code
- Feature extraction code
- Feature engineering
- Training scripts
- Saved model artifacts
- Model metadata
- Configuration/environment variables
- Prediction API endpoints
- Existing fallback logic
- Tests

Determine:

- Which models actually exist.
- Which models are trained.
- Which model artifacts are loaded at runtime.
- Where artifacts are stored.
- What feature columns each model expects.
- What feature columns the inference pipeline currently produces.
- Whether model metadata exists.
- What happens when an artifact is missing.
- What happens when an artifact is corrupted.
- What happens when the feature schema changes.

Do not assume that a model file existing in the repository means it is a valid production model.

---

# 2. MODEL ARTIFACT VALIDATION

Every model artifact used for prediction must be validated before inference.

Validate at minimum:

- Artifact exists.
- Artifact can be loaded.
- Artifact format is supported.
- Artifact is not corrupted.
- Model type is expected.
- Model version is known.
- Feature schema is known.
- Model metadata is available.
- Artifact is compatible with the current inference code.

If validation fails, the system must not claim that predictive ML intelligence is live.

---

# 3. MODEL METADATA

Every trained model should have associated metadata.

Where appropriate, include:

- model name
- model version
- model type
- trained timestamp
- training dataset/version
- training date range
- feature schema/version
- expected feature columns
- target variable
- evaluation metrics
- artifact identifier
- artifact checksum/hash where appropriate

Do not fabricate metadata.

If required metadata is unavailable, classify the model as incomplete/unvalidated rather than assuming it is production-ready.

---

# 4. FEATURE SCHEMA VALIDATION

The model's expected feature columns must match the actual input schema generated by the inference pipeline.

For every prediction request:

Model expected features
↓
Current feature generation
↓
Schema validation
↓
Inference

Validate:

- missing columns
- unexpected columns
- incorrect column names
- incorrect data types
- incorrect ordering where the model requires it
- incompatible categorical/numerical representation
- missing preprocessing steps

Do not silently rename or invent missing features unless that transformation is explicitly part of the model's preprocessing pipeline.

---

# 5. PREPROCESSING COMPATIBILITY

Verify that preprocessing used during training matches preprocessing used during inference.

Check:

- scaling
- encoding
- normalization
- categorical handling
- missing-value handling
- feature ordering
- feature transformations

Prefer storing the preprocessing pipeline together with the model where appropriate.

Do not load a model trained with one feature transformation and run it against a different transformation without explicit validation.

---

# 6. SCHEMA VERSIONING

Introduce a feature-schema version where appropriate.

For example:

`feature_schema_version = "v1"`

The model metadata should identify which schema version it expects.

At inference:

Current feature schema
must match
Model feature schema

If they do not match:

Reject the model for inference.

Do not silently continue.

---

# 7. MODEL PROVENANCE IN EVERY PREDICTION RESPONSE

Every prediction endpoint must expose model provenance and loaded status.

The response should contain appropriate metadata such as:

```json
{
  "prediction": {},
  "model": {
    "name": "risk_model",
    "version": "v1",
    "status": "loaded",
    "artifact": "validated",
    "feature_schema_version": "v1"
  }
}
````

Use the project's actual API response conventions.

Do not expose sensitive filesystem paths or internal secrets.

The important requirement is that the caller can determine whether the result came from:

* validated ML model
* fallback logic
* unavailable model

---

# 8. LOADED STATUS

Every prediction endpoint should clearly expose whether the intended model was successfully loaded.

Possible states:

* `LOADED`
* `UNAVAILABLE`
* `INVALID`
* `INCOMPATIBLE`
* `FALLBACK`

Do not report:

`LOADED`

unless the model was actually loaded and passed validation.

---

# 9. FALLBACK BEHAVIOR

The existing rule-based/statistical fallback may remain if it is useful for resilience.

However, it must be explicitly identified.

Correct behavior:

Valid model artifact
+
Compatible feature schema
+
Successful model loading
→ ML prediction

Missing/invalid model
→ Explicit fallback

The response must identify:

`prediction_mode = "ML"`

or:

`prediction_mode = "FALLBACK"`

Do not label fallback results as ML predictions.

---

# 10. PRODUCTION MODE

Introduce a production-safe configuration.

In production:

If predictive intelligence is required and the model artifact is:

* missing
* corrupted
* incompatible
* unvalidated
* schema-mismatched

the deployment/application must fail the predictive service readiness check.

Do NOT silently continue and claim that predictive ML intelligence is operational.

The exact behavior can be:

* application startup failure, or
* predictive-service readiness failure, or
* deployment health-check failure

depending on the existing architecture.

Do not unnecessarily prevent unrelated SAKSHA services from starting if the architecture supports isolated service readiness.

---

# 11. DEPLOYMENT VALIDATION

Add a deployment/startup validation step that verifies:

* required model artifacts exist
* artifacts load successfully
* model metadata exists
* feature schema matches
* preprocessing is compatible
* model version is valid

Production deployment should not be considered healthy if required predictive artifacts are missing or incompatible.

Do not make deployment depend on local developer filesystem paths.

---

# 12. MODEL CHECKSUM / INTEGRITY

Where practical, validate artifact integrity using a checksum/hash.

For example:

Expected artifact hash
vs
Actual artifact hash

If the artifact has unexpectedly changed or is corrupted:

Mark it invalid.

Do not invent checksums.

If the current deployment architecture does not support checksum management, implement a reasonable mechanism without overengineering.

---

# 13. RISK MODEL

Apply the validation requirements to:

`backend/app/ai/inference/risk.py`

Verify:

* correct artifact
* correct feature schema
* correct preprocessing
* model loading
* prediction mode
* model provenance
* error handling
* fallback behavior

The risk endpoint must not return a misleading ML prediction when the trained artifact is unavailable.

---

# 14. HOTSPOT MODEL

Apply the same requirements to:

`backend/app/ai/inference/hotspot.py`

Verify:

* correct artifact
* correct feature schema
* correct preprocessing
* model loading
* prediction mode
* model provenance
* error handling
* fallback behavior

Historical hotspot analysis must not be mislabeled as ML-based hotspot prediction.

---

# 15. MODEL REGISTRY / CENTRAL VALIDATION

If practical, create a reusable model validation/loading component rather than duplicating the same validation logic inside:

* `risk.py`
* `hotspot.py`

For example:

Model Registry
→ Locate artifact
→ Load metadata
→ Validate artifact
→ Validate schema
→ Load model
→ Expose model status

Then:

Risk inference
→ Validated risk model

Hotspot inference
→ Validated hotspot model

Follow the existing project architecture.

Do not introduce unnecessary complexity if a simpler shared service is sufficient.

---

# 16. FAIL SAFELY

If a model cannot be used, return a controlled response.

Do not expose:

* stack traces
* filesystem paths
* Python exceptions
* internal model details
* secrets

Example:

```text
Predictive model unavailable.
Validated ML inference cannot currently be performed.
```

If fallback is enabled:

```text
Validated ML model unavailable.
A statistical fallback was used.
```

The frontend should be able to distinguish these states.

---

# 17. DO NOT FABRICATE PERFORMANCE

Do not add fake:

* accuracy
* precision
* recall
* confidence
* model version
* training date
* validation score

If evaluation metadata does not exist, report it as unavailable.

The system must never create impressive-looking metrics simply to make the model appear production-ready.

---

# 18. MODEL FEATURE CONTRACT

Create a clear contract between training and inference.

The model should define:

Expected feature columns
+
Data types
+
Feature schema version
+
Preprocessing requirements

The inference pipeline must satisfy this contract before prediction.

If the contract is violated:

Prediction must be rejected.

---

# 19. TESTING

Create or extend tests covering:

### Test 1 — Valid Model

Valid artifact exists.

Expected:

Model loads successfully.

Prediction mode:

`ML`

---

### Test 2 — Missing Artifact

Artifact does not exist.

Expected:

Model status:

`UNAVAILABLE`

No misleading ML prediction.

---

### Test 3 — Corrupted Artifact

Artifact cannot be loaded.

Expected:

Model status:

`INVALID`

---

### Test 4 — Feature Mismatch

Model expects:

`feature_a`
`feature_b`
`feature_c`

Inference produces:

`feature_a`
`feature_b`

Expected:

Prediction rejected due to schema mismatch.

---

### Test 5 — Unexpected Feature

Inference contains an unsupported feature.

Expected:

Schema validation handles it according to the defined contract.

---

### Test 6 — Version Mismatch

Model expects feature schema v1.

Runtime provides v2.

Expected:

Model rejected as incompatible.

---

### Test 7 — Preprocessing Mismatch

Training and inference preprocessing configurations differ.

Expected:

Model rejected or handled according to the defined compatibility rules.

---

### Test 8 — Fallback

Valid ML model unavailable.

Expected:

Fallback may execute if enabled.

Response clearly identifies:

`prediction_mode = FALLBACK`

---

### Test 9 — Production Readiness

Production configuration requires ML model.

Artifact missing.

Expected:

Readiness/deployment validation fails.

---

### Test 10 — Provenance

Every successful prediction response contains:

* model name
* model version
* loaded status
* prediction mode
* feature schema version

where applicable.

---

# 20. FRONTEND COMPATIBILITY

Do not require major frontend redesign.

Ensure the existing prediction UI can consume the new metadata.

Where useful, display subtle information such as:

`ML Model v1`

or:

`Statistical Fallback`

or:

`Prediction Unavailable`

Do not expose technical filesystem/model details to normal users.

Use progressive disclosure for detailed model information.

---

# 21. REAL ENVIRONMENT VALIDATION

After implementation, test against the actual configured environment.

Verify:

1. Database is connected.
2. Feature generation works.
3. Required model artifact exists.
4. Model loads.
5. Feature schema matches.
6. Prediction executes.
7. API returns model provenance.
8. Frontend displays the correct prediction state.

Do not validate only with mocked model objects.

---

# 22. ACCEPTANCE CRITERIA

* [ ] Existing risk model pipeline has been audited.
* [ ] Existing hotspot model pipeline has been audited.
* [ ] Model artifacts are explicitly validated before inference.
* [ ] Model metadata is available and validated.
* [ ] Model and feature schema compatibility is checked.
* [ ] Preprocessing compatibility is checked.
* [ ] Feature schema versioning is implemented where appropriate.
* [ ] Every prediction endpoint exposes model provenance.
* [ ] Every prediction endpoint exposes loaded/model status.
* [ ] ML and fallback prediction modes are clearly distinguishable.
* [ ] Missing model artifacts cannot be silently presented as live ML intelligence.
* [ ] Invalid/corrupted artifacts are rejected.
* [ ] Feature mismatches are rejected.
* [ ] Production readiness checks validate required model artifacts.
* [ ] Production deployment/readiness fails when required artifacts are missing or incompatible.
* [ ] No fake model metrics or metadata are introduced.
* [ ] Model errors are handled safely.
* [ ] Automated tests cover artifact, schema, version, and fallback failures.
* [ ] Real environment validation has been performed.
* [ ] Existing risk/hotspot functionality remains intact.
* [ ] Existing SAKSHA functionality is not unnecessarily broken.

---

# Final Goal

The final predictive architecture should be:

Training Data
→ Feature Engineering
→ Versioned Feature Schema
→ Trained Model
→ Model Artifact
→ Artifact Validation
→ Schema Validation
→ Model Loading
→ Inference
→ Provenance Metadata
→ Prediction API
→ Frontend

The system must be able to answer:

* Which model generated this prediction?
* Was the model actually loaded?
* Which model version was used?
* Which feature schema was used?
* Was the artifact validated?
* Was the result generated by ML or fallback logic?
* Is the prediction service production-ready?

Most importantly:

A prediction must never be presented as validated live AI/ML intelligence unless a valid, compatible, trained model artifact is actually available and successfully loaded.

Do not remove the existing fallback unnecessarily.

Make the distinction between ML inference and fallback explicit, enforceable, testable, and visible to the system.

```
